File: backend/src/modules/transportation/dashboard/dashboard.controller.ts
import { Body, Controller, Get, Headers, Post, UseGuards } from '@nestjs/common';
import { HeaderAuthGuard, Validation } from '../../../common/auth/header-auth.guard';
import { CurrentUser } from '../../../common/auth/current-user.decorator';
import { DashboardService, AddAttendanceBody } from './dashboard.service';

// Base path /api/Transportation (faithful to the documented URLs, §6.9); every
// action requires the CompanyName + UserToken headers via the guard. Filters and
// pagination ride in HTTP headers (Node lowercases them), bodies are JSON.
@Controller('api/Transportation')
@UseGuards(HeaderAuthGuard)
export class DashboardController {
  constructor(private readonly service: DashboardService) {}

  @Get('DashBoard')
  dashboard(
    @Headers('dateserach') dateSearch?: string,
    @Headers('transportionlineid') lineId?: string,
    @Headers('supplierid') supplierId?: string,
    @Headers('serialbus') serialBus?: string,
    @Headers('routeid') routeId?: string,
  ) {
    return this.service.dashboard(dateSearch, {
      lineId: Number(lineId) || undefined,
      supplierId: Number(supplierId) || undefined,
      serialBus: serialBus || undefined,
      routeId: Number(routeId) || undefined,
    });
  }

  @Get('DashBoardForAttendenceDuration')
  dashboardAttendance(
    @Headers('fromdate') fromDate?: string,
    @Headers('todate') toDate?: string,
    @Headers('transportionlineid') lineId?: string,
    @Headers('supplierid') supplierId?: string,
    @Headers('serialbus') serialBus?: string,
    @Headers('routeid') routeId?: string,
    @Headers('attendaceflag') attendanceFlag?: string,
    @Headers('employeeid') employeeId?: string, // passenger/employee id (fingerprint no.)
    @Headers('pageno') pageNo?: string,
    @Headers('noofitems') noOfItems?: string,
  ) {
    return this.service.dashboardAttendance(Number(pageNo) || 1, Number(noOfItems) || 20, {
      fromDate: fromDate || undefined,
      toDate: toDate || undefined,
      lineId: Number(lineId) || undefined,
      supplierId: Number(supplierId) || undefined,
      serialBus: serialBus || undefined,
      routeId: Number(routeId) || undefined,
      attendanceFlag: attendanceFlag || undefined,
      employeeId: employeeId || undefined,
    });
  }

  @Get('AttendanceExcell')
  attendanceExcel(
    @Headers('fromdate') fromDate?: string,
    @Headers('todate') toDate?: string,
    @Headers('transportionlineid') lineId?: string,
    @Headers('supplierid') supplierId?: string,
    @Headers('serialbus') serialBus?: string,
    @Headers('routeid') routeId?: string,
    @Headers('employeeid') employeeId?: string,
  ) {
    return this.service.attendanceExcel({
      fromDate: fromDate || undefined,
      toDate: toDate || undefined,
      lineId: Number(lineId) || undefined,
      supplierId: Number(supplierId) || undefined,
      serialBus: serialBus || undefined,
      routeId: Number(routeId) || undefined,
      employeeId: employeeId || undefined,
    });
  }

  @Post('AddUsersAttedance')
  addUsersAttendance(@Body() body: AddAttendanceBody, @CurrentUser() v: Validation) {
    return this.service.addUsersAttendance(body, v.userId);
  }

  @Post('AddUsersAttedance44')
  addUsersAttendance44() {
    return this.service.addUsersAttendance44();
  }
}

File: backend-dotnet/GarasTransport.Api/Modules/Transportation/Dashboard/DashboardController.cs
using GarasTransport.Api.Common;
using Microsoft.AspNetCore.Mvc;

namespace GarasTransport.Api.Modules.Transportation.Dashboard;

// Base path /api/Transportation (§6.9). Every action requires CompanyName +
// UserToken headers via the filter.
[Route("api/Transportation")]
[HeaderAuth]
public class DashboardController : ApiControllerBase
{
    private readonly DashboardService _service;

    public DashboardController(DashboardService service) => _service = service;

    [HttpGet("DashBoard")]
    public Task<object> Dashboard() =>
        _service.Dashboard(Header("DateSerach"), new DashboardFilters
        {
            LineId = HeaderIntOrNull("TransportionLineId"),
            SupplierId = HeaderIntOrNull("SupplierId"),
            SerialBus = Header("SerialBus"),
            RouteId = HeaderIntOrNull("RouteId"),
        });

    private AttendanceFilters AttendanceFilters() => new()
    {
        FromDate = Header("FromDate"),
        ToDate = Header("ToDate"),
        LineId = HeaderIntOrNull("TransportionLineId"),
        SupplierId = HeaderIntOrNull("SupplierId"),
        SerialBus = Header("SerialBus"),
        RouteId = HeaderIntOrNull("RouteId"),
        AttendanceFlag = Header("AttendaceFlag"),
        EmployeeId = Header("EmployeeId"),
    };

    [HttpGet("DashBoardForAttendenceDuration")]
    public Task<object> DashboardAttendance() =>
        _service.DashboardAttendance(HeaderInt("PageNo", 1), HeaderInt("NoOfItems", 20), AttendanceFilters());

    [HttpGet("AttendanceExcell")]
    public Task<object> AttendanceExcel() => _service.AttendanceExcel(AttendanceFilters());

    [HttpPost("AddUsersAttedance")]
    public Task<object> AddUsersAttendance([FromBody] AddAttendanceBody body) =>
        _service.AddUsersAttendance(body ?? new(), CurrentUser.UserId);

    [HttpPost("AddUsersAttedance44")]
    public Task<object> AddUsersAttendance44() => _service.AddUsersAttendance44();
}
